### PR TITLE
Basic key generation

### PR DESCRIPTION
**Pull request by emmorais**
This PR adds basic key generation functionality.

### CODE REVIEW

File: eyelid-match-ops/Cargo.toml
Comment (teor2345) [inline eyelid-match-ops/Cargo.toml:49]: Production dependencies are automatically included in dev builds, so we can delete all 3 of these.

(I think we made then prod dependencies in previous PRs, but forgot to delete them here.)

File: eyelid-match-ops/src/primitives/poly.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:157]: If you import `ark_ff::UniformRand` into this module, you might be able to just do this:
```suggestion
        let coeff_rand = Coeff::rand(&mut rng);
```

Then there’s no need for any changes to `fq.rs`.

In general, we should be able to use `Coeff` everywhere, and never need to use any `Fq` types. Ticket #59 will make this easier.

**Review by teor2345 — COMMENTED** [review]
I wasn’t sure how much of a review you wanted here, so I just looked at the lowest-level functions.

Hopefully that’s enough to keep you going until we catch up later this week. Let me know if you have specific questions, or if you want to catch up after the team meeting.

File: eyelid-match-ops/src/primitives/poly.rs
Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly.rs:143]: These might be maths questions for @bgillesp as well?

Comment (bgillesp) [reply eyelid-match-ops/src/primitives/poly.rs:143]: Most important point: if the sampling is working correctly, integer overflow will not be a problem because the float values being sampled from the normal distribution should be small, i.e. at most one or two digits small, with high probability.  The probabilities decay exponentially with the magnitude of the output value.

Perhaps a radical suggestion, but it actually might make more sense for readability to use a _smaller_ datatype like an `i8` or `i16` for the float to integer conversion, to highlight the fact that the sampled values are expected to be small, and to make it clear that the conversion is not trying to faithfully represent a general `f64`.  In any case we should definitely be sure that the numbers sampled are distributed properly, which could either be its own check, or a test that the overall distribution of coefficients that comes out of the keygen procedure resembles the expected distribution.

Also wanted to mention, it looks like the casting from float to integer types in modern Rust (after 1.45.0) does essentially a clamp operation on values that are out of range -- but it turns out this was a compiler soundness issue until about five years ago: https://blog.rust-lang.org/2020/07/16/Rust-1.45.0.html#fixing-unsoundness-in-casts

Good question on the rounding mode; I think the detail of how the continuous normally distributed samples are to be converted to discrete integer coefficients is discussed in various places in the literature where key generation is addressed, but @emmorais is more familiar with the specific system he's implementing.  We should definitely double-check that the method implemented here matches the spec.

It might also be worth verifying that `Coeff::from` correctly handles signed integer values, but I would be surprised if this wasn't the case.

Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly.rs]: Yes, @bgillesp is right, we don't need big floats, on contrary, their size are limited, thus we could even use smaller types.

Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly.rs]: But the sampling indeed must be implemented carefully. I opened an [issue](https://github.com/Inversed-Tech/eyelid/issues/70) to review the distribution implementation, because it is a task that requires time and a different mindset. Since it can be done in parallel, we can proceed with the rest of the scheme implementation.

Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly.rs]: > Perhaps a radical suggestion, but it actually might make more sense for readability to use a _smaller_ datatype like an `i8` or `i16` for the float to integer conversion, to highlight the fact that the sampled values are expected to be small, and to make it clear that the conversion is not trying to faithfully represent a general `f64`. In any case we should definitely be sure that the numbers sampled are distributed properly, which could either be its own check, or a test that the overall distribution of coefficients that comes out of the keygen procedure resembles the expected distribution.

I would be fine with any type `u32` or smaller. But whatever the we choose, we should assert that the standard deviation is small enough to produce secure keys:
https://github.com/Inversed-Tech/eyelid/issues/70#issuecomment-2083687600

There’s also a significant probability edge case in generation, that we need to confirm isn’t discarding all the random bits (or if it does, the result is out of range and clamped):
https://github.com/Inversed-Tech/eyelid/issues/70#issuecomment-2083732955

This might also tighten the constraint on the standard deviation.

We can deal with this all in the ticket, but it seems important enough that I’ve added it to the milestone as a required security review.

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:100]: This seems like an important TODO that any external reviewers should know about:
```suggestion
            // TODO SECURITY: check that the generated integers are secure:
            // <https://github.com/Inversed-Tech/eyelid/issues/70>
            let normal = Normal::new(0.0, self.params.delta).unwrap();
```

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:44]: Are these `allow(dead_code)` still needed?

Usually they indicate that the type is unused _and_ not exported publicly. But these types should be public, if they're not that's a bug.

Or maybe there are some fields which are never read or written? That also seems like it could be a bug. Maybe the field could be deleted, or it needs a public accessor function?

```suggestion
/// Private key struct
#[derive(Debug, Clone)]
pub struct PrivateKey<const MAX_POLY_DEGREE: usize> {
    /// Sampled with small coefficients (and invertible)
    f: Poly<MAX_POLY_DEGREE>,
    /// The inverse of f
    finv: Poly<MAX_POLY_DEGREE>,
    /// Private key
    priv_key: Poly<MAX_POLY_DEGREE>,
}

/// Public key struct
#[derive(Debug)]
pub struct PublicKey<const MAX_POLY_DEGREE: usize> {
    /// Public key
    h: Poly<MAX_POLY_DEGREE>,
}
```

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:64]: Rust has a special syntax for this which is more performant than calling the potentially panicking function `unwrap()`:

```suggestion
            if let Ok(finv) = finv {
```

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:54]: If the RNG is broken or there's a bug in the code, key generation could become an infinite loop. (Or extremely slow.)

Should we loop a limited number of times instead, and return an error if that limit is exceeded?

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:15]: Should the type of this parameter be `Coeff` instead?
If we use `u64`, then the top 13 of the 77 bits can't be set (which also means we can't have small negative values).

Are there any security constraints on its value?
(For example, can it be zero or one?)

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:98]: If we gradually expand the polynomial, we will reallocate memory up to 11 times. This can cause performance issues.

If we make this method `pub(crate)`, we can use it here:
https://github.com/Inversed-Tech/eyelid/blob/facb2447c66dffe9b297cc729a8706b944fa9d84/eyelid-match-ops/src/primitives/poly/modular_poly.rs#L195-L197

We can make the same change in `sample_rand()`.

(This won't cause a panic, because we replace all the zeroes anyway, and truncate just to be sure.)

```rust
        let mut res = Poly::non_canonical_zeroes(MAX_POLY_DEGREE);
```

**Review by teor2345 — COMMENTED** [review]
Looks great!

Here is some feedback on the finished implementation. Some feedback is optional, or we can do it in other tickets.

File: eyelid-match-ops/src/primitives/poly.rs
Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly.rs]: This is blocking #59, so I opened ticket #71 and assigned it to myself.

Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly.rs]: Actually, I forgot I have to do invoices today, so I’ve un-assigned myself. Feel free to pick it up if you get time before Friday.

File: eyelid-match-ops/benches/match-ops.rs
Comment (emmorais) [reply eyelid-match-ops/benches/match-ops.rs:237]: It could be FULL_RES_POLY_DEGREE. I didn't find any bugs. 

The idea is to have a set of parameters for each resolution (high, medium, low). So MAX_POLY_DEGREE would be different for each one.

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (emmorais) [reply eyelid-match-ops/src/primitives/yashe.rs:97]: Thanks!

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (emmorais) [reply eyelid-match-ops/src/primitives/yashe.rs:16]: This parameter must be a positive integer greater than or equal to 2. It is the modulus of plaintext coefficients. For our instantiations, it will use at most 25 bits.

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (emmorais) [reply eyelid-match-ops/src/primitives/yashe.rs:53]: It is a high probability of getting good key material right in the first pass.

File: eyelid-match-ops/benches/match-ops.rs
Comment (emmorais) [reply eyelid-match-ops/benches/match-ops.rs]: Yes, the feedback is good. Sorry it happened again.

File: eyelid-match-ops/benches/match-ops.rs
Comment (teor2345) [inline eyelid-match-ops/benches/match-ops.rs:241]: Let's benchmark key generation as well.

This will need a change to the benchmark code, and an extra `|key` in the benchmark selection pattern here:
https://github.com/Inversed-Tech/eyelid/blob/facb2447c66dffe9b297cc729a8706b944fa9d84/.github/workflows/ci-bench-changes.yml#L55

File: eyelid-match-ops/benches/match-ops.rs
Comment (teor2345) [reply eyelid-match-ops/benches/match-ops.rs]: I think we can sort that out after ticket #59 gets merged. See the details here:
https://github.com/Inversed-Tech/eyelid/pull/69#discussion_r1583879731

(You might need to load more comments to see it.)

File: eyelid-match-ops/src/primitives/yashe/test/keygen.rs
Comment (teor2345) [reply eyelid-match-ops/src/primitives/yashe/test/keygen.rs:13]: This is now #73

File: eyelid-match-ops/benches/match-ops.rs
Comment (teor2345) [reply eyelid-match-ops/benches/match-ops.rs:241]: This is now ticket #72.

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (teor2345) [reply eyelid-match-ops/src/primitives/yashe.rs:56]: > It is a high probability of getting good key material right in the first pass.

What is the approximate probability? (order of magnitude is close enough)
That impacts how we deal with failures.

Did you want to deal with this in another ticket?
If so, we can just add another TODO in this PR, like this (you might need to load more comments to see it):
https://github.com/Inversed-Tech/eyelid/pull/69#discussion_r1583882681

Comment (teor2345) [reply eyelid-match-ops/src/primitives/yashe.rs:56]: This is now #74

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (teor2345) [reply eyelid-match-ops/src/primitives/yashe.rs:19]: Thanks, can you document that, and assert that it always holds? (if it's easy to add an assert)

I can deal with the types and any changes as part of #59.

File: eyelid-match-ops/src/primitives/poly/test/inv.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly/test/inv.rs:52]: Since this test is now testing key generation as well, if we get failures, it helps to rename it:

```suggestion
fn test_key_generation_and_inverse() {
```

File: eyelid-match-ops/src/primitives/poly.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:36]: Review note: I verified this code move using `git diff --color-moved main origin/ntru-keygen`

File: eyelid-match-ops/src/primitives/yashe/test/keygen.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/test/keygen.rs:40]: This assertion isn't guaranteed to hold due to leading zero coefficients.

For example:
https://github.com/Inversed-Tech/eyelid/actions/runs/8915222315/job/24484400937?pr=69#step:5:42

```suggestion
    assert!(public_key.h.degree() <= MAX_POLY_DEGREE - 1);
```

File: eyelid-match-ops/src/primitives/yashe/test/keygen.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/test/keygen.rs:40]:
```suggestion
    assert!(public_key.h.degree() < MAX_POLY_DEGREE);
```

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (teor2345) [reply eyelid-match-ops/src/primitives/yashe.rs:19]: I added the documentation task to ticket #70

File: eyelid-match-ops/src/primitives/yashe/test/keygen.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/test/keygen.rs:36]: The invertibility of the private key wasn't being checked during key generation:
https://github.com/Inversed-Tech/eyelid/actions/runs/8915386160/job/24484889593?pr=69#step:5:29

File: eyelid-match-ops/src/primitives/yashe.rs
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:63]:
```suggestion
            };
```

File: eyelid-match-ops/src/primitives/yashe/test/keygen.rs
Comment (teor2345) [reply eyelid-match-ops/src/primitives/yashe/test/keygen.rs:36]: Fixed by commit [ec3f328](https://github.com/Inversed-Tech/eyelid/pull/69/commits/ec3f32809671a4abc61076a464a942855d2de3c3)

**Review by teor2345 — APPROVED** [review]
This seems to be working, and the tests are passing after some small fixes.

**Pull request merged**
